File: lua/tst/mouse_warp.lua
require 'pico.check'

pico.init(true)

print("phy (500,500) -> log (50,50)")
do
    pico.set.window({title="Mouse Warp", dim={'!', w=500, h=500}})
    pico.set.view({dim={'!', w=50, h=50}})

    -- log (0,0) -> phy (0,0)
    do
        pico.set.mouse({'!', x=0, y=0, anchor='NW'})
        pico.input.event(100)
        local pos = pico.get.mouse()
        assert(pos.x==0 and pos.y==0)
    end

    -- pct (0.5, 0.5) -> log (25,25) -> phy (250,250)
    do
        pico.set.mouse({'%', x=0.5, y=0.5, anchor='NW'})
        pico.input.event(100)
        local pos = pico.get.mouse('%')
        assert(pos.x==0.5 and pos.y==0.5)
        local pos_raw = pico.get.mouse('!')
        assert(pos_raw.x==25 and pos_raw.y==25)
    end
end

-- Zoom out: src = {-25, -25, 100, 100}
-- centered 100x100 logical
print("zoom out 2x")
do
    pico.set.view({source={'!', x=-25, y=-25, w=100, h=100, anchor='NW'}})

    -- log (25, 25) -> phy (250, 250)
    do
        pico.set.mouse({'!', x=25, y=25, anchor='NW'})
        pico.input.event(100)
        local pos = pico.get.mouse()
        assert(pos.x==25 and pos.y==25)
    end

    -- log (-25, -25) -> phy (0, 0)
    do
        pico.set.mouse({'!', x=-25, y=-25, anchor='NW'})
        pico.input.event(100)
        local pos = pico.get.mouse()
        assert(pos.x==-25 and pos.y==-25)
    end
end

pico.init(false)

File: tst/mouse_warp.c
#include "pico.h"
#include <SDL2/SDL.h>
#include <stdio.h>
#include <assert.h>
#include <stdlib.h>

extern SDL_Window* pico_win;

int main(void) {
    pico_init(1);

    puts("phy (500,500) -> log (50,50)");
    {
        pico_set_window("Mouse Warp", -1, &(Pico_Rel_Dim){ '!', {500, 500}, NULL });
        pico_set_view(-1, &(Pico_Rel_Dim){ '!', {50, 50}, NULL }, NULL, NULL, NULL, NULL, NULL, NULL);

        // log (0,0) -> phy (0,0)
        {
            Pico_Rel_Pos p = {'!', {0, 0}, PICO_ANCHOR_NW, NULL};
            pico_warp_mouse(&p);
            SDL_PumpEvents();
            Pico_Mouse pos = pico_get_mouse(0);
            assert(pos.x==0 && pos.y==0);
        }

        // pct (0.5, 0.5) -> log (25,25) -> phy (250,250)
        {
            Pico_Rel_Pos p = {'%', {0.5, 0.5}, PICO_ANCHOR_NW, NULL};
            pico_warp_mouse(&p);
            SDL_PumpEvents();
            Pico_Mouse pos = pico_get_mouse('%');
            assert(pos.x==0.5 && pos.y==0.5);
            Pico_Mouse pos_raw = pico_get_mouse('!');
            assert(pos_raw.x==25 && pos_raw.y==25);
        }
    }

    // Zoom out: src = {-25, -25, 100, 100}
    // centered 100x100 logical
    puts("zoom out 2x");
    {
        pico_set_view(-1, NULL, NULL, NULL,
            &(Pico_Rel_Rect){ '!', {-25, -25, 100, 100}, PICO_ANCHOR_NW, NULL },
            NULL, NULL, NULL);

        // log (25, 25) -> phy (250, 250)
        {
            Pico_Rel_Pos p = {'!', {25, 25}, PICO_ANCHOR_NW, NULL};
            pico_warp_mouse(&p);
            SDL_PumpEvents();
            Pico_Mouse pos = pico_get_mouse(0);
            assert(pos.x==25 && pos.y==25);
        }

        // log (-25, -25) -> phy (0, 0)
        {
            Pico_Rel_Pos p = {'!', {-25, -25}, PICO_ANCHOR_NW, NULL};
            pico_warp_mouse(&p);
            SDL_PumpEvents();
            Pico_Mouse pos = pico_get_mouse(0);
            assert(pos.x==-25 && pos.y==-25);
        }
    }

    pico_init(0);
    return 0;
}
